linux_logging.md

```diff
@@ -0,0 +1,153 @@
+    - The are two systems in the Linux operating system that collects logs. `rsyslog` and `systemd-journal`.
+      - `rsyslog` makes permanent log in the `/var/log` directory
+        - Listens to the syslog socket for new logs.
+        - There are rules that controls which logs should be stored.
+      - `systemd-journal` keeps a temp log unless configured otherwise.
+        - A little weirder since the files are stored as binaries instead of a usually text file. Need to use `journalctl` to access the content.
+        - Stored files in the `/run/log/journal` directory.
+      - There is a way to store journal logs into rsyslog by changing the configuration of the rsyslog to take from journal.
+      - `logger` is a built in utility that will send strings to a syslog server.
+
+### Embedded readings
+
+`printf` is expensive, so logging should maybe be deferred.
+
+- An operating system called [Zephyr](https://docs.zephyrproject.org/latest/services/logging/index.html) uses deferred logging.
+- Move the logging operation elsewhere. The logging client does not need to concern itself with string manipulation since it might be responsible for time critical operations.
+
+The thing is, we are working with Linux, so we should be able to leverage the socket design paradigm to let the operating system do the scheduling work for us.
+
+## Syslog
+
+### Overview
+
+- `syslog` is a logging protocol, featuring log priority levels, message frame
+specifications, and logging resolution.
+- It runs on a single server daemon system, where a `syslog` daemon will
+be responsible for collecting all the `syslog` protocol logs that comes its
+way.
+- Most `syslog` daemons work with TCP, UDP, and UNIX sockets.
+
+### Note
+
+- Most Linux distros will have `rsyslog` come pre-installed.
+- `rsyslog` requires `systemd` to run.
+
+### Transaction protocol
+
+- Syslog uses UDP, usually at port 514, to pass message to the daemon.
+- It also has a TCP and Unix socket port that will accept `syslog` logs.
+
+### Syslog packet
+
+- Syslog uses UTF-8 encoding for its character.
+
+![Image of a syslog message packet](./assets/syslog_message_format.png)
+
+- Syslog packets are composed of *header*, *structured-data (SD)*, and a message.
+- **Header**:
+  - Priority
+  - Version
+  - Timestamp
+  - Hostname
+  - Application
+  - Process ID
+  - Message ID
+- **SD**:
+  - "key=value" format
+  - Unclear on usage for now
+- **Message**:
+  - Encoded using UTF-8
+  - Can be arbitarily long
+
+#### Example
+
+```
+<34>1 2003-10-11T22:14:15.003Z mymachine.example.com su - ID47 - BOM'su root' failed for lonvick on /dev/pts/8
+```
+
+Follows the following format:
+
+```
+<priority>VERSION ISOTIMESTAMP HOSTNAME APPLICATION PID MESSAGEID STRUCTURED-DATA MSG
+```
+
+### Log rotation
+
+- The syslog daemon will rotate log files out of use every week. Once the
+log file is two weeks old, the log file will get archived into a `.gz`.
```

Comment:
Is this configurable? We may want to store gzips of sys logs over more particular intervals like every 'N' orbits completed.

Reply:
Yes, in my reading there was a way to configure rsyslog.